internal/kmsg/kmsg.go

```diff
@@ -0,0 +1,149 @@
+// Package kmsg contains support for parsing Linux kernel log entries read from
+// /dev/kmsg. These are the same log entries that can be read via the `dmesg`
+// command. Each read from /dev/kmsg is guaranteed to return a single log entry,
+// so no line-splitting is required. If the read buffer is not large enough to
+// contain the log entry, `os.ErrInvalid` is returned.
+//
+// More information can be found here:
+// https://www.kernel.org/doc/Documentation/ABI/testing/dev-kmsg
+package kmsg
+
+import (
+	"errors"
+	"os"
+	"strconv"
+	"strings"
+	"syscall"
+
+	"github.com/sirupsen/logrus"
+)
+
+var (
+	// ErrInvalidFormat indicates the kmsg entry failed to parse.
+	ErrInvalidFormat = errors.New("invalid kmsg format")
+)
+
+// LogLevel represents the severity/priority of a log entry in the kernels
+// ring buffer.
+// https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/linux/kern_levels.h?id=HEAD
+type LogLevel uint8
+
+func (logLevel LogLevel) String() string {
+	levels := [...]string{
```

Comment:
nit: this syntax is new to me. Why isn't this just `[]string`?

Reply:
`[]string` will produce a slice, whereas `[...]string` will produce an array sized according to the number of elements used to initialize it. In this case, I don't think it matters.

Reply:
It would be interesting to look at the assembly.

Reply:
If I know the number of elements ahead of time I normally make an array but it really doesn't matter in this case. I wasn't aware of this syntax either, it was previously [8]string so I learned something new :)

Reply:
I looked at the assembly using `go build -gcflags -S file.go`. At a quick glance, it doesn't seem to make much difference between slice and array. What does seem to make a difference is whether the slice/array is a local or a global. It looks like the Go compiler will copy the slice/array onto the function's stack when you declare it inside like this.

So probably it's better to move this out as a global.

Reply:
That's kind of annoying. I would have hoped we would get that optimization if Go sees it doesn't need to really be function local.

I poked around a little at godbolt.org. I don't understand a lot of the assembly, but it looks like the array method avoids a heap allocation (the slice method has a call to `runtime.newobject`).

Reply:
I didn't see runtime.newobject in any of the 4 cases (slice vs array, globla vs local) when I ran it locally on Go 1.13.4.

Reply:
Interesting, could just be a platform difference. I'm using the "gc x86 1.13" configuration, which is presumably building for Linux.

Reply:
https://godbolt.org/z/F9jMR3

I don't see a way to set env vars via godbolt, unfortunately, so I can't change `GOOS`.

Reply:
Ah, I see. If you update the print calls to just grab one of the array/slice entries via indexing (a[i]) then you don't see the runtime.newobject call anymore.